File: botversion-sdk/cli/prompts.js
// botversion-sdk/cli/prompts.js

"use strict";

const readline = require("readline");

// ─── BASE PROMPT HELPER ───────────────────────────────────────────────────────

function ask(question) {
  const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout,
  });

  return new Promise((resolve) => {
    rl.question(question, (answer) => {
      rl.close();
      resolve(answer.trim());
    });
  });
}

function askChoice(question, choices) {
  const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout,
  });

  return new Promise((resolve) => {
    console.log("\n" + question);
    choices.forEach((choice, i) => {
      console.log(`  ${i + 1}. ${choice.label}`);
    });
    console.log("");

    const prompt = () => {
      rl.question(`Enter number (1-${choices.length}): `, (answer) => {
        const num = parseInt(answer.trim(), 10);
        if (num >= 1 && num <= choices.length) {
          rl.close();
          resolve(choices[num - 1]);
        } else {
          console.log(
            `  Please enter a number between 1 and ${choices.length}`,
          );
          prompt();
        }
      });
    };

    prompt();
  });
}

function confirm(question, defaultYes) {
  const hint = defaultYes ? "[Y/n]" : "[y/N]";
  return ask(`${question} ${hint}: `).then((answer) => {
    if (!answer) return defaultYes;
    return answer.toLowerCase().startsWith("y");
  });
}

// ─── SPECIFIC PROMPTS ─────────────────────────────────────────────────────────

async function promptMonorepoPackage(packages, cwd) {
  const path = require("path");
  const choices = packages.map((p) => ({
    label: path.relative(cwd, p),
    value: p,
  }));

  choices.push({ label: "This directory (root)", value: cwd });

  const choice = await askChoice(
    "We found multiple packages (monorepo). Which is your backend?",
    choices,
  );

  return choice.value;
}

async function promptEntryPoint() {
  console.log(
    "\n  ⚠️  We couldn't find your server entry point automatically.",
  );
  const filePath = await ask(
    "  Enter the path to your main server file (e.g. src/server.js): ",
  );
  return filePath;
}

async function promptAuthLibrary() {
  const choices = [
    {
      label: "next-auth v4",
      value: { name: "next-auth", version: "v4", supported: true },
    },
    {
      label: "next-auth v5 (Auth.js)",
      value: { name: "next-auth", version: "v5", supported: true },
    },
    { label: "Clerk", value: { name: "clerk", supported: true } },
    { label: "Supabase Auth", value: { name: "supabase", supported: true } },
    { label: "Passport.js", value: { name: "passport", supported: true } },
    { label: "JWT (jsonwebtoken)", value: { name: "jwt", supported: true } },
    {
      label: "express-session",
      value: { name: "express-session", supported: true },
    },
    { label: "Other / Custom", value: { name: "custom", supported: false } },
    { label: "No auth", value: { name: null, supported: false } },
  ];

  const choice = await askChoice(
    "We couldn't detect your auth library. Which one are you using?",
    choices,
  );

  return choice.value;
}

async function promptNextAuthConfigPath() {
  console.log(
    "\n  ⚠️  We couldn't find your authOptions location automatically.",
  );
  const filePath = await ask(
    "  Enter the path to your authOptions file (e.g. lib/auth.ts): ",
  );
  return filePath;
}

async function promptForce(conflictFile) {
  console.log(`\n  ⚠️  File already exists: ${conflictFile}`);
  return confirm("  Overwrite it? (a backup will be created)", false);
}

async function promptMissingListenCall(entryPoint) {
  console.log(`\n  ⚠️  We couldn't find app.listen() in ${entryPoint}`);
  console.log("  Options:");
  const choices = [
    { label: "Append to end of file", value: "append" },
    { label: "Enter the correct file path manually", value: "manual_path" },
    { label: "Skip — I'll add it manually", value: "skip" },
  ];

  const choice = await askChoice("How would you like to proceed?", choices);
  if (choice.value === "manual_path") {
    const filePath = await ask("  Enter file path: ");
    return { action: "manual_path", filePath };
  }

  return { action: choice.value };
}

module.exports = {
  ask,
  askChoice,
  confirm,
  promptMonorepoPackage,
  promptEntryPoint,
  promptAuthLibrary,
  promptNextAuthConfigPath,
  promptForce,
  promptMissingListenCall,
};

File: botversion-sdk-python/botversion_sdk/cli/prompts.py
"""
botversion-sdk-python/botversion-sdk/cli/prompts.py

Interactive CLI prompts — asks the user questions when things
can't be auto-detected.
Mirrors JS cli/prompts.js
"""


# ── Base prompt helpers ───────────────────────────────────────────────────────

def ask(question):
    """
    Asks a free-text question and returns the answer.
    Mirrors JS ask()
    """
    try:
        return input(question).strip()
    except (KeyboardInterrupt, EOFError):
        print("")
        raise SystemExit(0)


def ask_choice(question, choices):
    """
    Shows a numbered list and asks the user to pick one.
    Re-prompts if invalid input.
    Mirrors JS askChoice()
    """
    print(f"\n{question}")
    for i, choice in enumerate(choices):
        print(f"  {i + 1}. {choice['label']}")
    print("")

    while True:
        try:
            answer = input(f"Enter number (1-{len(choices)}): ").strip()
            num = int(answer)
            if 1 <= num <= len(choices):
                return choices[num - 1]
            else:
                print(f"  Please enter a number between 1 and {len(choices)}")
        except ValueError:
            print(f"  Please enter a number between 1 and {len(choices)}")
        except (KeyboardInterrupt, EOFError):
            print("")
            raise SystemExit(0)


def confirm(question, default_yes=True):
    """
    Asks a yes/no question.
    Mirrors JS confirm()
    """
    hint = "[Y/n]" if default_yes else "[y/N]"
    answer = ask(f"{question} {hint}: ")
    if not answer:
        return default_yes
    return answer.lower().startswith("y")


# ── Specific prompts ──────────────────────────────────────────────────────────

def prompt_entry_point():
    """
    Asks user to manually enter their server file path.
    Mirrors JS promptEntryPoint()
    """
    print("\n  ⚠  We couldn't find your server entry point automatically.")
    return ask("  Enter the path to your main server file (e.g. app.py or src/main.py): ")


# Config files that should never be appended to
CONFIG_FILES = [
    "settings.py",
    "config.py",
    "base.py",
    "development.py",
    "production.py",
    "wsgi.py",
    "asgi.py",
]

def prompt_missing_run_call(entry_point, framework):
    run_calls = {
        "flask": "app.run()",
        "fastapi": "uvicorn.run()",
        "django": "application = get_wsgi_application()",
    }
    run_call = run_calls.get(framework, "server start call")

    import os
    entry_filename = os.path.basename(entry_point or "")
    is_config_file = entry_filename in CONFIG_FILES

    print(f"\n  ⚠  We couldn't find {run_call} in {entry_point}")

    if is_config_file:
        print(f"\n  ❌  \"{entry_filename}\" is a config file, not a server file.")
        print("      Appending server code here would break your project.")
        print("  Options:")

        choices = [
            {"label": "Enter the correct server file path manually", "value": "manual_path"},
            {"label": "Skip — I'll add it manually", "value": "skip"},
        ]

        choice = ask_choice("How would you like to proceed?", choices)

        if choice["value"] == "manual_path":
            file_path = ask("  Enter file path: ")
            return {"action": "manual_path", "file_path": file_path}

        return {"action": choice["value"]}

    # Normal flow — not a config file
    print("  Options:")
    choices = [
        {"label": "Append to end of file", "value": "append"},
        {"label": "Enter the correct file path manually", "value": "manual_path"},
        {"label": "Skip — I'll add it manually", "value": "skip"},
    ]

    choice = ask_choice("How would you like to proceed?", choices)

    if choice["value"] == "manual_path":
        file_path = ask("  Enter file path: ")
        return {"action": "manual_path", "file_path": file_path}

    return {"action": choice["value"]}


def prompt_force(conflict_file):
    """
    Asks if user wants to overwrite an existing file.
    Mirrors JS promptForce()
    """
    print(f"\n  ⚠  File already exists: {conflict_file}")
    return confirm("  Overwrite it? (a backup will be created)", default_yes=False)


def prompt_django_settings_path():
    """
    Asks for the Django settings file path when auto-detection fails.
    """
    print("\n  ⚠  We couldn't find your Django settings file automatically.")
    return ask("  Enter the path to your settings file (e.g. config/settings.py): ")


def prompt_django_urls_path():
    """
    Asks for the Django urls.py path when auto-detection fails.
    """
    print("\n  ⚠  We couldn't find your main urls.py automatically.")
    return ask("  Enter the path to your main urls.py (e.g. myproject/urls.py): ")